File: tests/deepflow-python/app/utils/path_utils.py
"""Path utilities - HOP 15: Path operations.

Provides path manipulation that does NOT prevent path traversal.
"""

import os


def build_path(base: str, filename: str) -> str:
    """Build file path.

    HOP 15: DOES NOT sanitize path traversal.

    Args:
        base: Base directory path
        filename: Filename (TAINTED - may contain ..)

    Returns:
        Combined path (VULNERABLE to path traversal)
    """
    # VULNERABLE: os.path.join does NOT prevent path traversal
    # If filename starts with / or contains .., it can escape base
    # Example: filename = "../../etc/passwd"
    #          result = "/tmp/uploads/../../etc/passwd"
    #          which resolves to "/etc/passwd"
    return os.path.join(base, filename)


def get_extension(filename: str) -> str:
    """Get file extension.

    Args:
        filename: Filename (TAINTED)

    Returns:
        Extension (still part of TAINTED filename)
    """
    if not filename:
        return ""

    _, ext = os.path.splitext(filename)
    return ext.lower()


def normalize_path(path: str) -> str:
    """Normalize path (DOES NOT prevent traversal).

    This resolves .. and . but does NOT validate against base directory.

    Args:
        path: Path to normalize (TAINTED)

    Returns:
        Normalized path (still VULNERABLE)
    """
    # WARNING: normpath resolves .. but doesn't prevent traversal
    # It just makes the path cleaner
    return os.path.normpath(path)


def safe_join(base: str, *parts: str) -> str:
    """Safely join paths (SAFE VERSION).

    Used to demonstrate sanitized path detection.

    Args:
        base: Base directory (trusted)
        parts: Path parts to join (may be tainted)

    Returns:
        Safe joined path or raises ValueError if traversal detected

    Raises:
        ValueError: If path escapes base directory
    """
    # Join the paths
    path = os.path.join(base, *parts)

    # Resolve to absolute path
    resolved = os.path.realpath(path)
    base_resolved = os.path.realpath(base)

    # SAFE: Check that resolved path starts with base
    if not resolved.startswith(base_resolved + os.sep) and resolved != base_resolved:
        raise ValueError(f"Path traversal detected: {path}")

    return resolved

File: tests/deepflow-typescript/src/utils/path.utils.ts
/**
 * Path utilities - HOP 16: Path operations.
 *
 * Provides path manipulation that does NOT prevent path traversal.
 */

import * as path from 'path';
import * as fs from 'fs';

/**
 * Build file path.
 *
 * HOP 16: DOES NOT sanitize path traversal.
 *
 * @param base - Base directory path
 * @param filename - Filename (TAINTED - may contain ..)
 * @returns Combined path (VULNERABLE to path traversal)
 */
export function buildPath(base: string, filename: string): string {
  // VULNERABLE: path.join does NOT prevent path traversal
  // If filename contains .., it can escape base
  // Example: filename = "../../etc/passwd"
  return path.join(base, filename);
}

/**
 * Get file extension.
 *
 * @param filename - Filename (TAINTED)
 * @returns Extension (still part of TAINTED filename)
 */
export function getExtension(filename: string): string {
  if (!filename) return '';
  return path.extname(filename).toLowerCase();
}

/**
 * Normalize path (DOES NOT prevent traversal).
 *
 * @param filePath - Path to normalize (TAINTED)
 * @returns Normalized path (still VULNERABLE)
 */
export function normalizePath(filePath: string): string {
  // WARNING: normalize resolves .. but doesn't prevent traversal
  return path.normalize(filePath);
}

/**
 * Safely join paths (SAFE VERSION).
 *
 * @param base - Base directory (trusted)
 * @param parts - Path parts to join (may be tainted)
 * @returns Safe joined path or throws if traversal detected
 */
export function safeJoin(base: string, ...parts: string[]): string {
  // Join the paths
  const joined = path.join(base, ...parts);

  // Resolve to absolute path
  const resolved = path.resolve(joined);
  const baseResolved = path.resolve(base);

  // SAFE: Check that resolved path starts with base
  if (!resolved.startsWith(baseResolved + path.sep) && resolved !== baseResolved) {
    throw new Error(`Path traversal detected: ${joined}`);
  }

  return resolved;
}

/**
 * Read file (VULNERABLE).
 *
 * @param filePath - TAINTED file path
 */
export function readFile(filePath: string): string {
  // VULNERABLE: Reading from user-controlled path
  return fs.readFileSync(filePath, 'utf-8');
}
